gapis/api/vulkan/api/renderpass_framebuffer.api

```diff
@@ -163,100 +242,210 @@ cmd VkResult vkCreateRenderPass(
+        AspectMask: attachment.aspectMask
+      )
     }
+
     colorAttachments := subpass.pColorAttachments[0:subpass.colorAttachmentCount]
     for j in (0 .. subpass.colorAttachmentCount) {
-      description.ColorAttachments[j] = colorAttachments[j]
+      attachment := colorAttachments[j]
+      description.ColorAttachments[j] = AttachmentReference2(
+        Attachment: attachment.attachment,
+        Layout: attachment.layout,
+        AspectMask: attachment.aspectMask
+      )
     }
+
     if subpass.pResolveAttachments != null {
       resolveAttachments := subpass.pResolveAttachments[0:subpass.colorAttachmentCount]
       for j in (0 .. subpass.colorAttachmentCount) {
-        description.ResolveAttachments[j] = resolveAttachments[j]
+        attachment := resolveAttachments[j]
+        description.ResolveAttachments[j] = AttachmentReference2(
+          Attachment: attachment.attachment,
+          Layout: attachment.layout,
+          AspectMask: attachment.aspectMask
+        )
       }
     }
+
     if (subpass.pDepthStencilAttachment != null) {
-      depth_attachment := subpass.pDepthStencilAttachment[0]
-      description.DepthStencilAttachment = new!VkAttachmentReference(
-        Attachment: depth_attachment.Attachment,
-        Layout:     depth_attachment.Layout)
+      attachment := subpass.pDepthStencilAttachment[0]
+      description.DepthStencilAttachment = new!AttachmentReference2(
+        Attachment: attachment.attachment,
+        Layout: attachment.layout,
+        AspectMask: attachment.aspectMask,
+      )
     }
+
     preserveAttachments := subpass.pPreserveAttachments[0:subpass.preserveAttachmentCount]
     for j in (0 .. subpass.preserveAttachmentCount) {
       description.PreserveAttachments[j] = preserveAttachments[j]
     }
+
     renderPass.SubpassDescriptions[i] = description
   }
+
   dependencies := info.pDependencies[0:info.dependencyCount]
   for i in (0 .. info.dependencyCount) {
-    renderPass.SubpassDependencies[i] = dependencies[i]
+    dependency := dependencies[i]
+    renderPass.SubpassDependencies[i] = SubpassDependency2(
+      SrcSubpass: dependency.srcSubpass,
+      DstSubpass: dependency.dstSubpass,
+      SrcStageMask: dependency.srcStageMask,
+      DstStageMask: dependency.dstStageMask,
+      SrcAccessMask: dependency.srcAccessMask,
+      DstAccessMask: dependency.dstAccessMask,
+      DependencyFlags: dependency.dependencyFlags,
+      ViewOffset: dependency.viewOffset,
+    )
   }
-  handle := ?
-  if pRenderPass == null { vkErrorNullPointer("VkRenderPass") }
-  pRenderPass[0] = handle
-  renderPass.VulkanHandle = pRenderPass[0]
+
+  if info.pCorrelatedViewMasks != null {
+    pCorrelatedViewMasks := info.pCorrelatedViewMasks[0:info.correlatedViewMaskCount]
+    for i in (0 .. info.correlatedViewMaskCount) {
+      renderPass.CorrelatedViewMasks[i] = pCorrelatedViewMasks[i]
+    }
+  }
+
+  // We do not support any extension for this functionality yet.
```

Comment:
make sure pNext is null then, rather than just ignoring it.

Reply:
Well we affectively don't do that anywhere in the code.

This is our usual unhandled pNext code which can be found in many places:

```
if info.pNext != null {
    numPNext := numberOfPNext(info.pNext)
    next := MutableVoidPtr(as!void*(info.pNext))
    for i in (0 .. numPNext) {
      sType := as!const VkStructureType*(next.Ptr)[0]
      switch (sType) {
      }
      next.Ptr = as!VulkanStructHeader*(next.Ptr)[0].PNext
    }
```

This does not do anything but iterating over pNext while doing nothing. In reality, we shouldn't let any application with an unknown pNext to be traced and if they do we don't guarantee anything anyway.

Maybe it is actually a good idea to log or return error on an unknown pNext but that is out of the scope of this PR.

If you think that is necessary, we can go over every pNext and ensure that we do the proper handling when we do not support them.  Otherwise it would be inconsistent and confusing to do it for a particular function only.

And between and empty loop and no code  at all, I prefer 'no code'.

Reply:
@pmuetschard

Reply:
Melih, it does something. It goes through the whole chain and will observe at least the `sType` and `pNext` field of every struct during tracing. On mutate it will produce a nice error message containing which struct type we are not supporting, because the switch statement is missing a `default` case. We used to just ignore them, but that was bad. If you want you can add a comment inside the `switch`: `// Nothing supported yet, but ensures the pNext gets observed.`.

Reply:
I am sorry but it does not.

This is the generated code for those line(the function is `VkCreateQuery`) there is no error message here. As far as I know, we don't have nice error messages for trace at all.

```

if ((l_info.mpNext) != (nullptr)) {
                uint32_t l_numPNext = numberOfPNext(observer, l_info.mpNext);
                MutableVoidPtr l_next = MutableVoidPtr((void*)(l_info.mpNext));
                for (uint32_t l_i = static_cast<uint32_t>(0U); l_i < l_numPNext; ++l_i) {
                    uint32_t l_sType = observer->read(slice((uint32_t const*)(l_next.mPtr), static_cast<uint64_t>(0U), static_cast<uint64_t>(1U)), static_cast<uint64_t>(0U));
                    switch (l_sType) {
                    }
                    l_next.mPtr = observer->read(slice((VulkanStructHeader*)(l_next.mPtr), static_cast<uint64_t>(0U), static_cast<uint64_t>(1U)), static_cast<uint64_t>(0U)).mPNext;
                }
            }
```

Only in replay we do have a default case. But the way we do it, does not make any sense. No one reading that code can predict that we generate a default case(which we should not btw, we are generating something that contradicts with what's written) and we will do the error check there. There are much nicer ways to do this. For now, I will just check PNext and if it's not null, I'll return `vkErrUnsupported`

Reply:
Please check the pNext commit

Reply:
In our API language `switch` is an expression, not a statement, so it needs a "result value". I think in the C++ code generator the result value is the default value if none is specified, but in the Go generator it's considered a runtime-error to not have a value. This is because we don't really have abort/errors in C++ as we try to do our best to capture everything - trying to recover from error conditions, so you can inspect things when there are issues. On the GAPIS side we can be more strict and abort. Please don't be so quick to dismiss everything as bad or wrong, there is no contradiction here, things are just a bit more complex than you assumed. Of course there's always room to improve things.